### ISSUE TITLE
Unable to write headers / send HTTPS request 

### ISSUE THREAD

**Issue by joshuaballoch** [tuanpmt/esp-request#5]
Hi Tuan!

A coworker of mine has already written in about this, and I wanted to create an issue here so that we can all be involved in / kept up to date on the discussion.

We're getting an error trying to send an HTTPS request to an API we're running, with a 2048 bit certificate from LetsEncrypt. The error we're getting is:

```
E (8691) HTTP_REQ: Error write header
E (8691) HTTP_REQ: Error send request
```

For the record, you already mentioned in your email reply that you have also seen this behaviour.

Please let us know if there's anything you find on this... thanks!

**Comment by joshuaballoch** [comment]
Just another observation - we think it's just the encryption level of the key and not the certificate authority that is the issue. You had mentioned when you told us that you had run into this issue that it was also with a certificate issued by letsencrypt.

How we tested this:

1. We noticed that HTTPS requests to google.com worked with your library
2. We also noticed that google.com's key is 256 bits - much lower than the 2048 of the certificate we are using on our API.
3. We noticed that gmail.com's key is 2048 bits
4. We tried using the library to send a request to gmail.com
5. We also ran into an issue here.

Also, I notice in my nginx error log that we are able to see the initial handshake start and then the peer (the device on which we're running the request with your library) drops the connection before the SSL handshake is complete.

With these observations, we're pretty sure it's just an issue with the level of encryption and not the certificate authority or our specific encryption settings.

Were you able to find anything on your end?

**Comment by joshuaballoch** [comment]
FYI - when we debugged further, we found that the issue doesn't lie with your library.

Basically, we looked into your code, on the hunch that the issue with the SSL handshake was the inability to encrypt the pre-master shared secret. We notice that you just rely on the Open SSL library for this part, and looked deeper into that library.

Then we realized that our device was several versions behind on that (and other) libraries. We updated them and now we have no issue.

If you also still have this issue, this may be the same problem for you.

Thanks for your help!
